extensions/Khronos/KHR_materials_cmnConstant/README.md

```diff
@@ -0,0 +1,118 @@
+# KHR\_materials\_cmnConstant
+
+## Contributors
+
+* Don McCurdy, Google, <mailto:donmccurdy@google.com>
+
+## Status
+
+Draft (not ratified yet)
+
+## Dependencies
+
+Written against the glTF 2.0 spec.
+
+## Overview
+
+This extension defines an unlit ("constant") shading model for use in glTF 2.0
+materials, as an alternative to the Physically Based Rendering (PBR) shading
+models provided by the core specification. In mobile devices with limited
+resources, unlit materials offer a performant alternative to higher-quality
+shading models.
+
+## Extending Materials
+
+The common Constant material is defined by adding the
+`KHR_materials_cmnConstant` extension to any glTF material. When present, the
+extension indicates that a material should use the Constant shading model.
+Unrelated default values in the default PBR shading model, such as
+`roughnessFactor`, should be ignored.
+
+```
+{
+    "materials": [
+        {
+            "name": "MyUnlitMaterial",
+            "pbrMetallicRoughness": {
+                "baseColorFactor": [ 0.5, 0.8, 0.0, 1.0 ]
+            },
+            "extensions": {
+                "KHR_materials_cmnConstant": {}
+            }
+        }
+    ]
+}
+```
+
+### Definition
+
+The Constant material model describes a constantly shaded surface that is
+independent of lighting. The material is defined only by properties already
+present in the [glTF 2.0 material specification](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#material).
+No new properties are added by this extension — it is effectively a boolean
+flag indicating use of an unlit shading model. Additional properties on the
+extension object are allowed, but may lead to undefined behaviour in conforming
+viewers.
+
+Color is calculated as:
+
+```
+color = <emissiveTerm> + <baseColorTerm> * <occlusionTerm> * <aL>
+```
+
+Term definitions:
+
+* **emissiveTerm** – `emissiveFactor` * `emissiveTexture`
+    *  [*From core glTF material specification.*](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#additional-maps)
+* **occlusionTerm** – `occlusionTexture.strength` * `occlusionTexture`
+    *  [*From core glTF material specification.*](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#additional-maps)
+* **baseColorTerm** – `baseColorFactor` * `baseColorTexture` * `vertexColor`
+    *  [*From core glTF material specification.*](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#metallic-roughness-material)
+* **aL** – Indirect light
+    * *Constant sum of all ambient light contributions from the scene.*
```

Comment:
If we're saying that we're expecting baseFactor/baseTexture/vertexColor to define our unlit material (while tolerating emissive), then we shouldn't be multiplying it by **aL**, right? Unless the term's meaning here is meant to be entirely implicit.

Reply:
I second this, I don't think occlusion or ambient light should be part of this spec.

Reply:
Not sure I follow — are you asking whether an unlit material should ignore ambient light in addition to direct?

Reply:
Just saw @snagy's comment, the comment makes sense then —

On whether occlusion or ambient should be included, I'd point out that:

1. We did include them in the [constant material from glTF 1.0](https://github.com/KhronosGroup/glTF/tree/master/extensions/Khronos/KHR_materials_common#constant)
2. [THREE.MeshBasicMaterial supports both](https://threejs.org/docs/#api/materials/MeshBasicMaterial)
3. Disallowing ambient light has no particular performance benefit.

Is there a benefit to omitting occlusion and ambient light?

Reply:
It made sense to have an `aL` factor in the glTF 1.0 version because in that version we leaned on `emissive` to produce the unlit effect, and `emissive` is immune to scene light. That role is now to be played by baseColor & friends, instead. A material whose canonical visibility factor is multiplicatively attenuated by a light factor doesn't qualify as unlit, does it?

That is, unless we say that clients that do not have the concept of ambient light should assume a default value of 1.0... but that seems like a crazy can of worms...